guppylang-internals/src/guppylang_internals/definition/protocol.py

```diff
@@ -0,0 +1,149 @@
+)
+from guppylang_internals.diagnostic import Help
+from guppylang_internals.engine import DEF_STORE
+from guppylang_internals.error import GuppyError
+from guppylang_internals.span import SourceMap, Span, to_span
+from guppylang_internals.tys.arg import Argument
+from guppylang_internals.tys.param import Parameter, check_all_args
+from guppylang_internals.tys.protocol import ProtocolInst
+
+if TYPE_CHECKING:
+    from guppylang_internals.diagnostic import Error
+
+if sys.version_info >= (3, 12):
+    from guppylang_internals.tys.parsing import parse_parameter
+
+
+@dataclass(frozen=True)
+class ProtocolDef(Definition):
+    """Abstract base class for protocol definitions."""
+
+    description: str = field(default="protocol", init=False)
+
+
+@dataclass(frozen=True)
+class DeclarationHint(Help):
+    message: ClassVar[str] = (
+        "Add a `@guppy.declare` annotation and leave the body empty to turn this into "
+        "a declaration"
+    )
+
+
+@dataclass(frozen=True)
+class RawProtocolDef(ProtocolDef, ParsableDef):
+    """A raw protocol definition that has not been parsed yet."""
+
+    python_class: type
+
+    def parse(self, globals: Globals, sources: SourceMap) -> "CheckedProtocolDef":
+        """Parses the raw class object into an AST and checks that it is well-formed."""
+        # Mostly copied from RawStructDef.parse, but only allowing function declarations
+        # in the body and allowing `Protocol` as a base class.
+        frame = DEF_STORE.frames[self.id]
+        cls_def = parse_py_class(self.python_class, frame, sources)
+        if cls_def.keywords:
+            raise GuppyError(UnexpectedError(cls_def.keywords[0], "keyword"))
+
+        # Look for generic parameters from Python 3.12 style syntax.
+        params = []
+        params_span: Span | None = None
+        if sys.version_info >= (3, 12):
+            if cls_def.type_params:
+                first, last = cls_def.type_params[0], cls_def.type_params[-1]
+                params_span = Span(to_span(first).start, to_span(last).end)
+                params = [
+                    parse_parameter(node, idx, globals)
+                    for idx, node in enumerate(cls_def.type_params)
+                ]
+
+        match cls_def.bases:
+            case []:
+                pass
+            # We allow `Generic[...]` to specify  parameters with the legacy syntax.
+            case [base] if elems := try_parse_generic_base(base):
+                # Complain if we already have Python 3.12 generic params
+                if params_span is not None:
+                    err: Error = RedundantParamsError(base, self.name)
+                    err.add_sub_diagnostic(RedundantParamsError.PrevSpec(params_span))
+                    raise GuppyError(err)
+                params = params_from_ast(elems, globals)
+            # Specifying `Protocol` is redundant but we allow it optionally.
+            case [base] if isinstance(base, ast.Name) and base.id == "Protocol":
```

Comment:
Python also allows specifying legacy generic parameters via a `Protocol[T]` base: https://typing.python.org/en/latest/spec/protocol.html#generic-protocols

Do we also want this?